accidentally typing `cabal (non-command) <path>` gives bogus output instead of reporting that 'non-command' does not exist

minimal reproducible example:

```bash
$ git clone git@github.com:chessai/semirings && cd semirings
$ cabal new-sdist
$ cabal new-upload dist-newstyle/sdist/semirings-0.2.1.1.tar.gz
Resolving dependencies...
Build profile: -w ghc-8.6.1 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (configuration changed)
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..

Main.hs:1:1: error: lexical error at character '\US'
  |
1 | �������}kw�Ȳ��̯�}��6�`;N�3�
  | ^

```

Here's a picture:

![what](https://user-images.githubusercontent.com/18648043/46310277-f0274980-c58c-11e8-9d36-1d7a91212611.png)

Compare the above to the following, which behaves correctly:

```bash
$ cabal new-upload dist-newstyle/sdist/semirings-0.2.1.1.tar.gz --publish
cabal: unrecognised command: new-upload (try --help)
```

The same behavior can also be triggered by passing an unknown flag to an existing command.
And you don't even need a .cabal file.
```
$ mkdir tempdir
$ cd tempdir
$ touch some-file
$ cabal new-clean --foo some-file
Resolving dependencies...
Build profile: -w ghc-8.4.3 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (first run)
Configuring executable 'script' for fake-package-0..
Preprocessing executable 'script' for fake-package-0..
Building executable 'script' for fake-package-0..
[1 of 1] Compiling Main             ( Main.hs, /tmp/cabal-repl.-7293/dist-newstyle/build/x86_64-linux/ghc-8.4.3/fake-package-0/x/script/build/script/script-tmp/Main.o )

Main.hs:1:1: error:
    Could not find module ‘Prelude’
    Use -v to see a list of the files searched for.
```

It seems this happens when there should be an error in parsing the arguments (unknown subcommand or unknown flag) and the last argument is a filename that exists.

```
$ cabal new-clean --foo non-existant-file
cabal: unrecognized 'new-clean' option `--foo'
$ cabal new-clean some-file --foo
cabal: unrecognized 'new-clean' option `--foo'
$ cabal new-clea non-existant-file
cabal: unrecognised command: new-clea (try --help)
$ cabal new-clea some-file
Resolving dependencies...
Build profile: -w ghc-8.4.3 -O1
In order, the following will be built (use -v for more details):
 - fake-package-0 (exe:script) (first run)
Configuring executable 'script' for fake-package-0..
...
```

--- a/cabal-install/Distribution/Client/CmdRun.hs
+++ b/cabal-install/Distribution/Client/CmdRun.hs
@@ -9,7 +9,7 @@ module Distribution.Client.CmdRun (
     -- * The @run@ CLI and action
     runCommand,
     runAction,
-    handleShebang,
+    handleShebang, validScript,
 
     -- * Internals exposed for testing
     TargetProblem(..),
@@ -104,7 +104,7 @@ import qualified Text.Parsec as P
 import System.Directory
          ( getTemporaryDirectory, removeDirectoryRecursive, doesFileExist )
 import System.FilePath
-         ( (</>) )
+         ( (</>), isValid, isPathSeparator )
 
 
 runCommand :: CommandUI (ConfigFlags, ConfigExFlags, InstallFlags, HaddockFlags, TestFlags)
@@ -298,9 +298,31 @@ runAction (configFlags, configExFlags, installFlags, haddockFlags, testFlags)
                   installFlags haddockFlags testFlags
     globalConfigFlag = projectConfigConfigFile (projectConfigShared cliConfig)
 
-handleShebang :: String -> IO ()
-handleShebang script =
-  runAction (commandDefaultFlags runCommand) [script] defaultGlobalFlags
+-- | Used by the main CLI parser as heuristic to decide whether @cabal@ was
+-- invoked as a script interpreter, i.e. via
+--
+-- > #! /usr/bin/env cabal
+--
+-- or
+--
+-- > #! /usr/bin/cabal
+--
+-- As the first argument passed to `cabal` will be a filepath to the
+-- script to be interpreted.
+--
+-- See also 'handleShebang'
+validScript :: String -> IO Bool
+validScript script
+  | isValid script && any isPathSeparator script = doesFileExist script
+  | otherwise = return False
+
+-- | Handle @cabal@ invoked as script interpreter, see also 'validScript'
+--
+-- First argument is the 'FilePath' to the script to be executed; second
+-- argument is a list of arguments to be passed to the script.
+handleShebang :: FilePath -> [String] -> IO ()
+handleShebang script args =
+  runAction (commandDefaultFlags runCommand) (script:args) defaultGlobalFlags
 
 parseScriptBlock :: BS.ByteString -> ParseResult Executable
 parseScriptBlock str =
@@ -366,14 +388,14 @@ handleScriptCase verbosity baseCtx tempDir scriptContents = do
   -- We need to create a dummy package that lives in our dummy project.
   let
     sourcePackage = SourcePackage
-      { packageInfoId        = pkgId
-      , SP.packageDescription   = genericPackageDescription
-      , packageSource        = LocalUnpackedPackage tempDir
-      , packageDescrOverride = Nothing
+      { packageInfoId         = pkgId
+      , SP.packageDescription = genericPackageDescription
+      , packageSource         = LocalUnpackedPackage tempDir
+      , packageDescrOverride  = Nothing
       }
-    genericPackageDescription = emptyGenericPackageDescription 
+    genericPackageDescription  = emptyGenericPackageDescription
       { GPD.packageDescription = packageDescription
-      , condExecutables    = [("script", CondNode executable' targetBuildDepends [])]
+      , condExecutables        = [("script", CondNode executable' targetBuildDepends [])]
       }
     executable' = executable
       { modulePath = "Main.hs"

--- a/cabal-install/main/Main.hs
+++ b/cabal-install/main/Main.hs
@@ -186,7 +186,7 @@ import qualified Paths_cabal_install (version)
 import System.Environment       (getArgs, getProgName)
 import System.Exit              (exitFailure, exitSuccess)
 import System.FilePath          ( dropExtension, splitExtension
-                                , takeExtension, (</>), (<.>))
+                                , takeExtension, (</>), (<.>) )
 import System.IO                ( BufferMode(LineBuffering), hSetBuffering
                                 , stderr, stdout )
 import System.Directory         (doesFileExist, getCurrentDirectory)
@@ -234,10 +234,9 @@ main' = do
 
 mainWorker :: [String] -> IO ()
 mainWorker args = do
-  validScript <-
-    if null args
-      then return False
-      else doesFileExist (last args)
+  hasScript <- if not (null args)
+    then CmdRun.validScript (head args)
+    else return False
 
   topHandler $
     case commandsRun (globalCommand commands) commands args of
@@ -253,7 +252,7 @@ mainWorker args = do
           CommandHelp     help           -> printCommandHelp help
           CommandList     opts           -> printOptionsList opts
           CommandErrors   errs
-            | validScript                -> CmdRun.handleShebang (last args)
+            | hasScript                  -> CmdRun.handleShebang (head args) (tail args)
             | otherwise                  -> printErrors errs
           CommandReadyToGo action        -> do
             globalFlags' <- updateSandboxConfigFileFlag globalFlags

